src/index_notation/transformations.cpp

```diff
@@ -131,6 +135,73 @@ std::ostream& operator<<(std::ostream& os, const Reorder& reorder) {
   return os;
 }
 
+// class SetMergeStrategy
+
+struct SetMergeStrategy::Content {
+  IndexVar i_var;
+  MergeStrategy strategy;
+};
+
+SetMergeStrategy::SetMergeStrategy(IndexVar i, MergeStrategy strategy) : content(new Content) {
+  content->i_var = i;
+  content->strategy = strategy;
+}
+
+IndexVar SetMergeStrategy::geti() const {
+  return content->i_var;
+}
+
+MergeStrategy SetMergeStrategy::getMergeStrategy() const {
+  return content->strategy;
+}
+
+IndexStmt SetMergeStrategy::apply(IndexStmt stmt, string* reason) const {
+  INIT_REASON(reason);
+
+  string r;
+  if (!isConcreteNotation(stmt, &r)) {
+    *reason = "The index statement is not valid concrete index notation: " + r;
+    return IndexStmt();
+  }
+
+  struct SetMergeStrategyRewriter : public IndexNotationRewriter {
+    using IndexNotationRewriter::visit;
+
+    SetMergeStrategy transformation;
+    string* reason;
+    SetMergeStrategyRewriter(SetMergeStrategy transformation, string* reason)
+            : transformation(transformation), reason(reason) {}
+
+    IndexStmt setmergestrategy(IndexStmt stmt) {
+      return rewrite(stmt);
+    }
+
+    void visit(const ForallNode* node) {
+      Forall foralli(node);
+      IndexVar i = transformation.geti();
+
+      if (foralli.getIndexVar() == i) {
```

Comment:
Since you already have the `string*` for error checking output here, you should add verification of the preconditions you write earlier in `index_notation.h`. You can construct a merge lattice here to verify the conditions you laid out earlier -- see `Parallelize::apply` for an example.